File: android/src/main/java/ru/innim/flutter_login_vk/LoginCallback.java
package ru.innim.flutter_login_vk;

import com.vk.api.sdk.auth.VKAccessToken;
import com.vk.api.sdk.auth.VKAuthCallback;

import java.util.HashMap;

import io.flutter.plugin.common.MethodChannel;

public class LoginCallback implements VKAuthCallback {
    private MethodChannel.Result _pendingResult;

    public void addPending(MethodChannel.Result result) {
        if (_pendingResult != null)
            callError(FlutterError.interrupted("Interrupted by another login call", null));

        _pendingResult = result;
    }

    @Override
    public void onLogin(final VKAccessToken accessToken) {
        VKClient.token = accessToken;
        callResult(Results.loginSuccess(accessToken));
    }

    @Override
    public void onLoginFailed(int errorCode) {
        if (errorCode == VKAuthCallback.AUTH_CANCELED) {
            callResult(Results.loginCancelled());
        } else {
            callError(FlutterError.apiError("Login failed: " + errorCode,
                    new VKError(errorCode, null)));
        }
    }

    private void callResult(HashMap<String, Object> data) {
        if (_pendingResult != null) {
            _pendingResult.success(data);
            _pendingResult = null;
        }
    }

    private void callError(FlutterError error) {
        if (_pendingResult != null) {
            _pendingResult.error(error.code, error.message, error.details);
            _pendingResult = null;
        }
    }
}

File: android/src/main/kotlin/ru/innim/flutter_login_vk/LoginCallback.kt
package ru.innim.flutter_login_vk

import com.vk.api.sdk.auth.VKAccessToken
import com.vk.api.sdk.auth.VKAuthCallback
import com.vk.api.sdk.exceptions.VKAuthException
import io.flutter.plugin.common.MethodChannel
import java.util.HashMap

class LoginCallback : VKAuthCallback {
    private var pendingResult: MethodChannel.Result? = null

    fun addPending(result: MethodChannel.Result) {
        if (pendingResult != null) {
            callError(FlutterError.interrupted("Interrupted by another login call"))
        }
        pendingResult = result
    }

    override fun onLogin(token: VKAccessToken) {
        callResult(Results.loginSuccess(token))
    }

    override fun onLoginFailed(authException: VKAuthException) {
        if (authException.isCanceled) {
            callResult(Results.loginCancelled())
        } else {
            val errorCode = authException.webViewError
            callError(FlutterError.apiError("Login failed: $errorCode",
                    VKError(errorCode, authException.authError)))
        }
    }

    private fun callResult(data: HashMap<String, Any>) {
        pendingResult?.success(data)
        pendingResult = null
    }

    private fun callError(error: FlutterError) {
        pendingResult?.error(error.code, error.message, error.details)
        pendingResult = null
    }
}